### PR TITLE
Smart search for tagging: prioritize relevant users in @ mentions

### LINKED ISSUE #4280: Smart search for tagging
user exmateriae suggests:
> when you're trying to @ someone, it works in really weird ways once you type letter and not the @ alone. It does not seem to be related at all to how often I tag people which is a bummer because my issue is often with someone specific : [@johnnycaffeine](https://www.metaculus.com/accounts/profile/115975), I swear I might have spent an hour by now with how often I tag you and your name never pops up until I've written "johnnyc", that makes no sense to me. I make that even worse because I regularly forget where the h is or how many n there are

This would require the implementation of a smart search that takes into account the user's previous tags. We don't currently have the infrastructure to address this quickly, but it might be do-able.

My initial thought would be to create a table of tags where you have the tagger, taggee, and number of times tagged. Then, when a user starts the tag (no letters typed yet), the endpoint it pings will first return that list in order of number of times tagged. Each time a letter is typed we apply that filter but still sort in order of number of times tagged. Seems like a good idea, but might not be worth the implementation time.
One of the biggest annoyances with this would be going through all of the existing comments to generate that tag-table. Of course, we can just start it fresh and ignore previous comments.

### PR DESCRIPTION
**Pull request by SylvainChevalier**
Closes #4280

## Summary
- Add optional `post_id` param to `/users/` search endpoint to prioritize relevant users (commenters, authors, project members) in @ mention suggestions
- Thread `postId` through frontend mentions context to API call
- Add backend tests for relevance-based user search

## Test plan
- [ ] Run `python -m pytest tests/unit/test_users/test_search.py -v`
- [ ] Test @ mentions in comment editor on a post with known commenters
- [ ] Verify that commenters appear first when searching by username
- [ ] Verify existing @ mention behavior still works without post context

Generated with [Claude Code](https://claude.ai/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Mentions and search can optionally target a specific post so suggestions prioritize commenters, authors, coauthors, and permission holders.

* **Behavior Changes**
  * User search responses are returned as a flat, non-paginated list limited to 20 items.
  * Backend relevance ordering is preserved when a post filter is used.
  * Searches must include either a search term or a post filter.
  * User list endpoint now requires authentication.

* **Tests**
  * Added comprehensive tests for post-filtered relevance, ordering, exclusions, and edge cases.

### CODE REVIEW

**Comment by SylvainChevalier** [comment]
@claude keep the requirement of 3 characters before suggestions.

**Comment by SylvainChevalier** [comment]
@claude I know that, I'm asking who receives notifications when @ members is used.

File: front_end/src/services/api/profile/profile.shared.ts [hunk 1]
```diff
@@ -7,9 +7,14 @@ class ProfileApi extends ApiService {
...
+    const params = new URLSearchParams();
+    params.set("search", query);
+    if (postId) {
+      params.set("post_id", String(postId));
+    }
     return await this.get<PaginatedPayload<UserProfile>>(
-      `/users/?search=${query}`
+      `/users/?${params.toString()}`
     );
   }
```
Comment (hlbmtc) [inline front_end/src/services/api/profile/profile.shared.ts:16]: Utilize `encodeQueryParams`

File: front_end/src/components/markdown_editor/plugins/mentions/utils.ts [hunk 2]
```diff
@@ -7,13 +7,10 @@ export async function queryMentions(
...
-  userPermission?: ProjectPermissions
+  userPermission?: ProjectPermissions,
+  postId?: number
 ): Promise<MentionItem[]> {
-  const usersGroupMentions = [
-    { value: "moderators" },
-    { value: "admins" },
-    { value: "members" },
-  ];
+  const usersGroupMentions = [{ value: "moderators" }, { value: "admins" }];
```
Comment (hlbmtc) [inline front_end/src/components/markdown_editor/plugins/mentions/utils.ts:13]: Why did we deprecate "members" here?

File: users/services/common.py [hunk 3]
```diff
@@ -23,26 +24,118 @@
...
+    When post_id is provided, users relevant to that post are prioritized:
+    - Users who have commented on the post
+    - The post author and coauthors
+    - Users with explicit permissions on the post's default project
+
+    Non-priority users are filtered to only those who are active and have
+    posted a non-deleted comment in the last year.
     """
+    from comments.models import Comment
+    from posts.models import Post
```
Comment (hlbmtc) [inline users/services/common.py]: Remove inline import

File: users/services/common.py [hunk 4]
```diff
@@ -23,26 +24,118 @@
...
+
+    one_year_ago = timezone.now() - timezone.timedelta(days=365)
+
+    # User IDs with at least one non-deleted comment in the last year
+    recently_active_user_ids = (
+        Comment.objects.filter(
+            is_soft_deleted=False,
+            created_at__gte=one_year_ago,
```
Comment (hlbmtc) [inline users/services/common.py]: Do we have a db index?

File: users/services/common.py [hunk 5]
```diff
@@ -23,26 +24,118 @@
...
+            full_match=Case(
+                When(username__iexact=search, then=1),
+                default=0,
+                output_field=IntegerField(),
+            )
+        ).filter(username__icontains=search)
+
+    # Annotate relevance when post_id is provided
+    if post_id:
+        post = Post.objects.filter(pk=post_id).first()
```
Comment (hlbmtc) [inline users/services/common.py]: Ensure user has permissions of this post not to expose its members

File: users/services/common.py [hunk 6]
```diff
@@ -23,26 +24,118 @@
...
+    Non-priority users are filtered to only those who are active and have
+    posted a non-deleted comment in the last year.
     """
+    from comments.models import Comment
+    from posts.models import Post
+
+    one_year_ago = timezone.now() - timezone.timedelta(days=365)
+
+    # User IDs with at least one non-deleted comment in the last year
+    recently_active_user_ids = (
```
Comment (hlbmtc) [inline users/services/common.py]: Please note: this endpoint is triggered almost on every char input, so we should care about performance. 
For this specific function, move it to the separate construction and cache it -- we don't really need to refresh it live every time

File: users/services/common.py [hunk 7]
```diff
@@ -23,26 +24,118 @@
 
 def get_users(
```
Comment (hlbmtc) [inline users/services/common.py:49]: For `users_list_api_view`:
- Remove pagination - I don't think we really need it there
- Limit response to 20 records

File: users/services/common.py [hunk 8]
```diff
@@ -20,31 +24,146 @@
 from utils.email import send_email_with_template
 from utils.frontend import build_frontend_email_change_url
 
+RECENTLY_ACTIVE_USERS_CACHE_KEY = "recently_active_user_ids"
+RECENTLY_ACTIVE_USERS_CACHE_TIMEOUT = 60 * 60  # 1 hour
+
+
+def get_recently_active_user_ids() -> set[int]:
```
Comment (hlbmtc) [inline users/services/common.py:33]: For caching: use `cached_singleton` util

File: users/services/common.py [hunk 9]
```diff
@@ -20,31 +24,146 @@
...
+                default=0,
+                output_field=IntegerField(),
+            )
+        ).filter(username__icontains=search)
+
+    # Annotate relevance when post_id is provided
+    if post_id:
+        # Verify the requesting user has permission to view this post
+        post_qs = Post.objects.annotate_user_permission(user)
+        post = post_qs.filter(pk=post_id).first()
```
Comment (hlbmtc) [inline users/services/common.py]: Replace with:

```
    permission = get_post_permission_for_user(post, user=request.user)
    ObjectPermission.can_view(permission, raise_exception=True)
```

File: users/services/common.py [hunk 10]
```diff
@@ -20,31 +24,146 @@
...
+                )
+
+            return list(qs[:20])
+
+    if search:
+        # Without post_id, only return recently active users
+        qs = qs.filter(id__in=recently_active_user_ids)
+        qs = qs.order_by("-full_match", "username")
+
+    return list(qs[:20])
```
Comment (hlbmtc) [inline users/services/common.py]: This function should return QuerySet! Make cut in the caller view

File: tests/unit/test_users/test_search.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    @pytest.fixture(autouse=True)
+    def setup(self):
+        self.url = reverse("users-list")
+
+    def _make_recently_active(self, user):
+        """
+        Give the user a recent non-deleted comment so they pass the activity filter.
+        """
+        other_post = factory_post(author=factory_user())
+        factory_comment(author=user, on_post=other_post)
```
[the remark this answers is not in the document]
Comment (hlbmtc) [reply tests/unit/test_users/test_search.py:34]: @claude

File: users/services/common.py [hunk 12]
```diff
@@ -17,33 +23,137 @@
...
             )
-            .filter(username__icontains=search)
-            .order_by("-full_match", "username")
+        ).filter(username__icontains=search)
+
+    # Annotate relevance when post_id is provided
+    if post_id:
+        post = Post.objects.filter(pk=post_id).first()
+        if not post:
+            return User.objects.none()
```
Comment (hlbmtc) [inline users/services/common.py]: Replace with `get_object_or_404`

File: users/services/common.py [hunk 13]
```diff
@@ -17,33 +24,135 @@
...
+        # Verify the requesting user has permission to view this post
+        permission = get_post_permission_for_user(post, user=user)
+        ObjectPermission.can_view(permission, raise_exception=True)
+
+        # Collect user IDs of commenters on this post
+        commenter_ids = (
+            Comment.objects.filter(on_post_id=post_id)
+            .values_list("author_id", flat=True)
+            .distinct()
         )
```
[the remark this answers is not in the document]
Comment (hlbmtc) [reply users/services/common.py:93]: @claude

**Comment by hlbmtc** [comment]
@claude take care of this nits:
> 68-68: get_recently_active_user_ids() is eagerly fetched even when it won't be used.
When post_id is provided without search, the function returns at Line 149 without ever referencing recently_active_user_ids. Moving this call closer to its actual usage points (Lines 129 and 153) would avoid a cache lookup + potential DB query in that code path.

> The is_authenticated guard on Line 96 is dead code given DRF defaults.
Since the AllowAny decorator was removed, this view now uses the project-wide IsAuthenticated default permission class. request.user is guaranteed to be authenticated when this code runs, so the else None branch is unreachable. Consider simplifying:

**Comment by hlbmtc** [comment]
@claude:

- I've just made /api/users/ endpoint for auth users only. Adjust tests
- take care of this nits:
> 68-68: get_recently_active_user_ids() is eagerly fetched even when it won't be used.
When post_id is provided without search, the function returns at Line 149 without ever referencing recently_active_user_ids. Moving this call closer to its actual usage points (Lines 129 and 153) would avoid a cache lookup + potential DB query in that code path.

> The is_authenticated guard on Line 96 is dead code given DRF defaults.
Since the AllowAny decorator was removed, this view now uses the project-wide IsAuthenticated default permission class. request.user is guaranteed to be authenticated when this code runs, so the else None branch is unreachable. Consider simplifying:
